db/crud.go

```diff
@@ -627,29 +627,40 @@ func (db *Database) PutExistingRev(docid string, body Body, docHistory []string)
 // Common subroutine of Put and PutExistingRev: a shell that loads the document, lets the caller
 // make changes to it in a callback and supply a new body, then saves the body and document.
 func (db *Database) updateDoc(docid string, allowImport bool, expiry uint32, callback func(*document) (Body, AttachmentData, error)) (newRevID string, err error) {
-	_, newRevID, err = db.updateAndReturnDoc(docid, allowImport, expiry, callback)
+	_, newRevID, err = db.updateAndReturnDoc(docid, allowImport, expiry, nil, nil, 0, callback)
 	return newRevID, err
 }
 
-func (db *Database) updateAndReturnDoc(docid string, allowImport bool, expiry uint32, callback func(*document) (Body, AttachmentData, error)) (docOut *document, newRevID string, err error) {
+// Calling updateAndReturnDoc directly allows callers to:
+//   1. Receive the updated document body in the response
+//   2. Specify the existing document body/xattr/cas, to avoid initial retrieval of the doc in cases that the current contents are already known (e.g. import).
+//      On cas failure, the document will still be reloaded from the bucket as usual.
+func (db *Database) updateAndReturnDoc(
+	docid string,
+	allowImport bool,
+	expiry uint32,
+	existingValue []byte, // If existingValue, existingCas, and existingXattr are present, passes these to WriteUpdateWithXattr to allow bypass of initial GET
+	existingXattr []byte,
+	existingCas uint64,
+	callback func(*document) (Body, AttachmentData, error)) (docOut *document, newRevID string, err error) {
```

Comment:
I think newlines after the params definitely make this easier to read, and at some point we should create a named type for that callback which will help a lot.